### ISSUE TITLE
LFX Mentorship (Jun-Aug, 2024): Finetune LLM models for Rust coding assistance

### ISSUE THREAD

**Issue by juntao** [WasmEdge/WasmEdge#3371]
## Summary

WasmEdge is a lightweight inference runtime for AI and LLM applications. We want to build specialized and finetuned models for WasmEdge community. The model should be supported by WasmEdge and its applications should benefit the WasmEdge community. 

In this project, we will build and compare two finetuned model for Rust coding assistance.

* A code review model. It aims to be a new backend for the [PR review bot](https://github.com/flows-network/github-pr-summary/) we currently use in the community.
* A QA model. It should be able to answer user questions about the Rust language and provide explanations. Our goal is to provide an alternative to our [Learn Rust](https://flows.network/learn-rust) app.

### Objective 1: Code review model

**Create a dataset with the following two fields** 

We are looking for at least 200 Q and A pairs. The total length of each QA should be less than 3000 words.

```
Q: a code segment
A: explanation / review of the code
```

The QA could come from Rust documentation such as [Rust by Example](https://doc.rust-lang.org/rust-by-example/) and [The Rust Programming Language](https://doc.rust-lang.org/book/).

**Assemble the dataset into the llama3 chat template** 

It is similar to the following. Each entry should be all in one line with linebreaks denoted as `\n`.

```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

You are a reviewer of Rust source code.<|eot_id|><|start_header_id|>user<|end_header_id|>

{{ a code segment }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>

{{ explanation / review of the code }}<|eot_id|>
```

**Finetune** 

we will finetune based on the [llama3-8b-instruct model](https://huggingface.co/second-state/Llama-3-8B-Instruct-GGUF). 

You are free to use any finetune tools. But if you are unsure, we recommend using llama.cpp's `finetune` utility. [See an example](https://github.com/YuanTony/chemistry-assistant/tree/main/fine-tune-model). It can run on CPUs. We will provide the computing resources required for the finetuing.

### Objective 2: Code QA model

**Create a dataset with the following three fields**

We are looking for at least 100 chapter + Q + A rows.

```
C: A chapter from a Rust book
Q: A question related to the chapter
A: Explanation / answer for the question
```

You could use ChatGPT to generate these questions and answers based on the chapter content.

**Assemble the dataset into the llama3 chat template**

It is similar to the following. Each entry should be all in one line with linebreaks denoted as \n.

```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

You are an expert of the Rust language. Please answer question based on the context below.\n---------\n{{ book chapter }}<|eot_id|><|start_header_id|>user<|end_header_id|>

{{ a code segment }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>

{{ explanation / review of the code }}<|eot_id|>
```

**Finetune**

Due to the chapter-long context length used in this dataset, we will finetune based on the [262k long context length llama3-8b-instruct model](https://huggingface.co/second-state/Llama-3-8B-Instruct-262k-GGUF). 

You are free to use any finetune tools. But if you are unsure, we recommend using llama.cpp's finetune utility. [See an example](https://github.com/YuanTony/chemistry-assistant/tree/main/fine-tune-model). It can run on CPUs. We will provide the computing resources required for the finetuing.

### Objective 3: Compare the two finetuned models

Start the finetuned models using the [LlamaEdge API server](https://github.com/LlamaEdge/LlamaEdge/tree/main/api-server), and test them on commonly used scenarios.

## LFX

Expected outcome: Two finetuned models based on Llama3-8b for Rust code review and QA.

Recommended skills:
- Rust language
- ChatGPT and Claude
- LlamaEdge
- llama.cpp

Mentor:
- Michael Yuan @juntao michael@secondstate.io

Application link: [ttps://mentorship.lfx.linuxfoundation.org/project/d52d172e-598d-4817-be97-3338d5b96432](https://mentorship.lfx.linuxfoundation.org/project/d52d172e-598d-4817-be97-3338d5b96432)

### Appendix

* [Rust by Example](https://doc.rust-lang.org/rust-by-example/) 
* [The Rust Programming Language](https://doc.rust-lang.org/book/)

**Comment by DhruvSinghiitmandi** [comment]
Hi @juntao!
I am Dhruv , Currently pursuing my undergrad in CS from IIT Mandi.
I am deeply interested in this mentorship project and I have some experience in creating dataset for finetuning llms and finetuning llms. Recently I did a [project](https://github.com/DhruvSinghiitmandi/Health.ai) where i created medical Conversation dataset using ddxplus dataset using GPT-3.5 for further finetuning a medllama2(llama2 finetuned on medQA dataset) on a single RTX-3090 using QLora. 

Could you please let me know if there's a pretest or any other steps I should take to participate?

**Comment by debrupf2946** [comment]
@juntao 
I am Debrup pursuing Engg at BITS Pilani, and a contributor at Keras, I like to solve NLP problems and have worked extensively with hugginface(transformers, PEFT, datasets), Langchain, Llama index(data injection and Indexing), Knowledge graphs(Neo4J), RAGs, Rust.
I was also selected for GSoC in a similar project building coding assistant using Knowledge Graphs, which assist with QA and Summarization tasks of any Github Repo,
I find this project interesting and similar, and I want to contribute, and take part in the LFX-Mentorship program
![gsoc_pr](https://github.com/WasmEdge/WasmEdge/assets/126772439/95e9a611-edaa-48c4-8a39-17fe510536c9)

**Comment by debrupf2946** [comment]
@juntao Is there any community(Discord, slack..) where we contributors can interact? What are the timelines and selection criteria for this project I am really excited about the project?
If possible can share a cover letter for the previous term I wanted to get an idea of length,content.

**Comment by angad-singhh** [comment]
Hey @juntao | @hydai 
I'm Angad Singh, a final year grad student. I have gone through all 4 - LFX mentorship project, which WasmEdge is participating and I find this project most interesting and relevant to my skills and experience. 

I will go through all the resources , links mentioned above and update you both with my progress on the community channel as well as here to improve my chances of getting selected for the LFX program.

**Comment by angad-singhh** [comment]
### Progress till 15th May 2024:

I have started working and exploring the first project objective that is: 

> A code review model. It aims to be a new backend for the [PR review bot](https://github.com/flows-network/github-pr-summary/) we currently use in the community.

### What I plan to do:

1. Start with exploring the current [PR review bot](https://github.com/flows-network/github-pr-summary/) used in the community and why we need to improve this. 
2. Then I try setting up the current [PR review bot](https://github.com/flows-network/github-pr-summary/), onto my own GitHub account
3. Analyse the prior PR submitted on the WasmEdge repo and understand how the current [PR review bot](https://github.com/flows-network/github-pr-summary/) responds in all PR.

### What all I did:

I was able to complete all the planned steps I have mentioned above, below are my findings till now about the current PR review Bot.

**Findings** | please verify this part (@juntao @hydai )

- First thing I noticed is that WasmEdge currently use flow template github-pr-summary written in Rust language and is hosted on the WasmEdge runtime on flows.network. There is also a flow template named github-pr-review with slightly different working.
- The working of current PR bot (github-pr-summary) is to collect the patches of commits from the PR and send it to the OpenAI to summarize is commit and afterward it combines all these summaries into one single summary and sends back the combine summary of all commits + summary of each commits in the PR
- Thing to note is that, it automatically updates the PR review comment whenever a new commit is been made in the PR and appends the detail summary/review of that commit. 

### Issues 

1. Although I was able to set up the [PR review bot](https://github.com/flows-network/github-pr-summary/) bot onto my own GitHub account. But there was a small issue was regarding OpenAI API key authentication, for which I have a raised a detailed issue [here](https://github.com/flows-network/docs/issues/39).
2. While I was exploring the previous PR's in the WasmEdge repo, I noticed that the [PR review bot](https://github.com/flows-network/github-pr-summary/) was not giving any output in all PR's made after the March 2024, like in these PR [1](https://github.com/WasmEdge/WasmEdge/pull/3404), [2](https://github.com/WasmEdge/WasmEdge/pull/3396),[3](https://github.com/WasmEdge/WasmEdge/pull/3386), ....

**flows.network setup**
<img width="300" alt="Screenshot 2024-05-15 013841" src="https://github.com/WasmEdge/WasmEdge/assets/108326021/4aa8bab2-d9bb-4748-983d-35f226071457">

**Comment by angad-singhh** [comment]
### Next 

I'm currently exploring the Fine-tuning part, I will try to setup the fine-tuned chemistry assistant [here](https://github.com/YuanTony/chemistry-assistant/tree/main/fine-tune-model) which you provided and make some changes and understand its working. 

Also, I will be exploring the llama.cpp model, its working , docs, to get better understand and update you all with any further progress. 

Lastly, I will try to look for any _good first issues_ in code/docs and work around it

**Comment by codemaster1104** [comment]
Hello @juntao @hydai , myself Akshat Shrivastava in past few days i have understood the workflow of finetuning and running llm's locally, i was wondering weather for current task i could use llama-3-8b-bnb-4bit as its a direct modification of llama3 8b but in 4 bit to run everything in my current hardware smoothly

**Comment by angad-singhh** [comment]
> Hello @juntao @hydai , myself Akshat Shrivastava sophomore at IIT BHU(Varanasi) in past few days i have understood the workflow of finetuning and running llm's locally, i was wondering weather for current task i could use llama-3-8b-bnb-4bit as its a direct modification of llama3 8b but in 4 bit to run everything in my current hardware smoothly
Hey @codemaster1104 ,
It has been mentioned in the project issue that we are supposed to use [Llama-3-8B-Instruct-GGUF](https://huggingface.co/second-state/Llama-3-8B-Instruct-GGUF) for fine-tuning. 

For the tool, llama.cpp is preferred one, because of its capability to run on CPU itself, also the resources required for fine-tuning will be provide by Organization itself. 

I hope I cleared some of your doubts !!

**Comment by angad-singhh** [comment]
### Progress Till 17th May 2024:

1. I was able to successfully set up and run the required  [llama-3-8B-Instruct-GGUF](https://huggingface.co/second-state/Llama-3-8B-Instruct-GGUF) for this project on my local machine using the LlamaEdge environment.
2. Next I also went through the llamaEdge repo, llamaEdge API server and its working.
3. I also found and issue in [WasmEdge docs](https://github.com/WasmEdge/docs/issues/218), discussed here in the comments.

**llama-3-8b-Instruct-GGUF -test setup** 

><img width="853" alt="Screenshot 2024-05-17 180758" src="https://github.com/WasmEdge/WasmEdge/assets/108326021/7976bab6-9fe3-4bce-b9b8-17b70c770dd1">
>
>![Screenshot (442)](https://github.com/WasmEdge/WasmEdge/assets/108326021/10ecc9a2-a785-49e4-a645-ac024a789e97)

**Comment by angad-singhh** [comment]
### Next up: 

1. I have started working on the Data-set for QA model
2. I have started revising my Rust concepts
3. I will try to understand the llama.cpp workflow and its working with a small project.

**Comment by codemaster1104** [comment]
![Screenshot from 2024-05-18 01-59-41](https://github.com/WasmEdge/WasmEdge/assets/115527374/384f607d-4417-48b3-a63a-f2c7b2875b44)
Till now i have made a small dataset , and have successfully converted to required prompt template , example:

['You are a reviewer of Rust source code.\n\n\n### Code:\nprintln!({} days, 31);\n\n### Explanation:\nPrints 31 days using {} as a placeholder for the value (31).<|end_of_text|>', 'You are a reviewer of Rust source code.\n\n\n### Code:\nprintln!({0}, this is {1}. {1}, this is {0}, Alice, Bob);\n\n### Explanation:\nPrints Alice, this is Bob. Bob, this is Alice using positional arguments ({0}, {1}). Order matters!<|end_of_text|>', 'You are a reviewer of Rust source code.\n\n\n### Code:\nprintln!({subject} {verb} {object}, subject=the lazy dog, verb=jumps over, object=quick brown fox);\n\n### Explanation:\nPrints a sentence with named arguments, improving readability.<|end_of_text|>', "You are a reviewer of Rust source code.\n\n\n### Code:\nlet logical: bool = true;\n\n### Explanation:\nDeclares a boolean variable named 'logical' and initializes it with the value 'true'.<|end_of_text|>", "You are a reviewer of Rust source code.\n\n\n### Code:\nlet a_float: f64 = 1.0;\n\n### Explanation:\nDeclares a floating-point variable named 'a_float' and initializes it with the value '1.0'. The type 'f64' represents a 64-bit floating-point number.<|end_of_text|>",

Tokenizer is also working fine:
{'input_ids': tensor([[128000,   2675,    527,    264,  56614,    315,  34889,   2592,   2082,
           4286,  14711,   6247,    512,  34755,      0,   2358,     92,   2919,
             11,    220,   2148,    629,  14711,  72387,    512,   9171,     82,
            220,   2148,   2919,   1701,   4792,    439,    264,   6002,    369,
            279,    907,    320,   2148,    570]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}
         
   Next will be working on training the model with current dataset and then will move to improve my dataset

**Comment by codemaster1104** [comment]
![Screenshot from 2024-05-18 14-13-37](https://github.com/WasmEdge/WasmEdge/assets/115527374/da187c49-3904-4840-b3a2-cdac2339961e)
I successfully fine tuned llama3-8b-Instruct model with my small dataset, the attached image is of training loss which is decreasing steadily which is a good sign, next will be working on creating the final dataset and fine tuning

@juntao could you please guide us with the metrics we could use to compare models  , currently i am assuming that we would be manually comparing a bunch of answers for specific questions in both models, but i am sure there would be a better way

**Comment by codemaster1104** [comment]
![ll2](https://github.com/WasmEdge/WasmEdge/assets/115527374/f7ab79db-9280-4110-9352-e61600d3d627)
![mm2](https://github.com/WasmEdge/WasmEdge/assets/115527374/4a15fcf9-e46e-4faf-9f0e-d2f5479937ca)
These are the direct comparison between vanilla llama3-8b-Instruct and my fine tuned model for Code Explanation, My model appears to be much more direct and on point which was expected as my dataset of training is similarly direct and on point

@juntao @hydai really need your guidance here is such direct behaviour desired?, according to me for pr review bot should give on point explanations, If this behaviour is desired i will start working on my final dataset with similar types of explanations to code.

**Comment by codemaster1104** [comment]
![Screenshot from 2024-05-22 17-11-11](https://github.com/WasmEdge/WasmEdge/assets/115527374/a8eccd23-d068-4d05-b155-8186a98d2462)
Got llamaedge up and running and inference via chat endpoint for my fine tuned model is working fine

**Comment by pawaspy** [comment]
Hey @juntao I have keen interest in this project. I have good hands-on experience with Large language models and made several project similar to this QNA project, which utilizes the OpenAI API and answers the latest answers. I would really be glad to in this project and refining my knowledge in this project. Could you tell me how I could interact with the members and other people.?

**Comment by codemaster1104** [comment]
![Frame 2](https://github.com/WasmEdge/WasmEdge/assets/115527374/37e36645-9a90-46cf-b869-1596e2f35293)
There are my results for now for fine tuned model of objective 1 and objective 2, both of these models are running on LlamaEdge API server , for model of objective 2 I improved on dataset and added content of entire chapter in chapter field , hence some of my fields reached more than 50000 words and to fine tune it i had to reduce context length to 131072 and change other parameters from recommended value to avoid running into out of RAM issue.

My model for objective 2 is still hallucinating and it seems to get stuck on the context of previous questions, I would like to fine tune this model with recommended parameters first for which i will need more computing resources, I have also started to read papers on the topic to get better understanding.

# My progress

- Created dataset for both objectives( can be viewed on my hugging face given in my cover letter)
- Did initial fine tuning (Collab for same can be found on my cover letter)
- Did some experimentation with parameters and compared the result 
- Improved some instructions for LllamaEdge API server and my PR got merged : [Link](https://github.com/LlamaEdge/LlamaEdge/pull/169) 
- Fine tuned with different iterations of dataset to test how it changes the performance

**Comment by staru09** [comment]
Is there any test planned for this project?

**Comment by codemaster1104** [comment]
@juntao Thanks for selecting me as a mentee for this issue, how do you want me to get started with the issue?, and how should i structure my tasks, should i create a new PR/Issue or just communicate about progress here

**Comment by hydai** [comment]
> @juntao Thanks for selecting me as a mentee for this issue, how do you want me to get started with the issue?, and how should i structure my tasks, should i create a new PR/Issue or just communicate about progress here

Please raise an issue such as [this one](https://github.com/WasmEdge/WasmEdge/issues/2545) to describe the mission, milestones, and timeline. All of the communication should be done on it for tracking. Thanks.

**Comment by codemaster1104** [comment]
I have created an [issue](https://github.com/WasmEdge/WasmEdge/issues/3495) , currently I have created a rough timeline for myself which I will be improving according to mentor reviews
